features/mbedtls/targets/TARGET_STM/aes_alt.c

```diff
@@ -147,31 +150,93 @@ int mbedtls_aes_crypt_ecb( mbedtls_aes_context *ctx,
 
 #if defined(MBEDTLS_CIPHER_MODE_CBC)
 #if defined (TARGET_STM32L486xG)
+static int st_cbc_restore_context(mbedtls_aes_context *ctx){
+    uint32_t tickstart;
+    tickstart = HAL_GetTick();
+    while((ctx->hcryp_aes.Instance->SR & AES_SR_BUSY) != 0){
+        if ((HAL_GetTick() - tickstart) > ST_AES_TIMEOUT) {
+            return 1; // timeout: CRYP processor is busy
+        }
+    }
+    /* allow multi-instance of CRYP use: restore context for CRYP hw module */
+    ctx->hcryp_aes.Instance->CR = ctx->ctx_save_cr;
+    return 0;
+}
+static int st_cbc_save_context(mbedtls_aes_context *ctx){
+    uint32_t tickstart;
+    
+    tickstart = HAL_GetTick();
+    while((ctx->hcryp_aes.Instance->SR & AES_SR_BUSY) != 0){
+        if ((HAL_GetTick() - tickstart) > ST_AES_TIMEOUT) {
+            return 1; // timeout: CRYP processor is busy
+        }
+    }
+    /* allow multi-instance of CRYP use: save context for CRYP HW module CR */
+    ctx->ctx_save_cr = ctx->hcryp_aes.Instance->CR;
+
+    return 0;
+}
 static int st_hal_cryp_cbc( mbedtls_aes_context *ctx, uint32_t opmode, size_t length, 
                             unsigned char iv[16], uint8_t *input, uint8_t *output) 
 {
     int status = 0;
     ctx->hcryp_aes.Init.pInitVect = &iv[0]; // used in process, not in the init
-    if ((ctx->hcryp_aes.Init.OperatingMode != opmode) || \
-         (ctx->hcryp_aes.Init.ChainingMode != CRYP_CHAINMODE_AES_CBC) || \
-         (ctx->hcryp_aes.Init.KeyWriteFlag != CRYP_KEY_WRITE_ENABLE)) {
-
-        /* Re-initialize AES IP with proper parameters */
-        if (HAL_CRYP_DeInit(&ctx->hcryp_aes) != HAL_OK)
-            return HAL_ERROR;
-        ctx->hcryp_aes.Init.OperatingMode = opmode;
-        ctx->hcryp_aes.Init.ChainingMode = CRYP_CHAINMODE_AES_CBC;
-        ctx->hcryp_aes.Init.KeyWriteFlag = CRYP_KEY_WRITE_ENABLE;
-        if (HAL_CRYP_Init(&ctx->hcryp_aes) != HAL_OK)
-            return HAL_ERROR;
-    }
+    /* At this moment only, we know we have CBC mode: Re-initialize AES
+        IP with proper parameters and apply key and IV for multi context usecase */
+    if (HAL_CRYP_DeInit(&ctx->hcryp_aes) != HAL_OK)
+        return HAL_ERROR;
+    ctx->hcryp_aes.Init.OperatingMode = opmode;
+    ctx->hcryp_aes.Init.ChainingMode = CRYP_CHAINMODE_AES_CBC;
+    ctx->hcryp_aes.Init.KeyWriteFlag = CRYP_KEY_WRITE_ENABLE;
+    if (HAL_CRYP_Init(&ctx->hcryp_aes) != HAL_OK)
+        return HAL_ERROR;
 
     status =  HAL_CRYPEx_AES(&ctx->hcryp_aes, input, length, output, 10);
 
     return status;
 }
+#else
```

Comment:
Could we document why we need two different implementations for save/restore depending on the target family?